keps/993-two-phase-admission/README.md

```diff
@@ -0,0 +1,274 @@
+[...]
+	// Status of admission checks, if there are any specified 
+	// for the ClusterQueue in which the workload is queued.
+	AdmissionChecksConditions []metav1.Condition `json:"admissionChecksConditions,omitempty"`
+}
+
+
+// Cluster-scoped top-level object defining a check that 
+// can be referenced in ClusterQueue
+type AdmissionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec AdmissionCheckSpec `json:"spec,omitempty"`
+}
+
+const (
+	// The ckech cannot pass at this moment, back off (possibly 
+	// allowing other to try, unblock quota) and retry.
+	Retry metav1.ConditionStatus = "Retry"
+	// The check will not pass in the near future. It is not worth
+	// to retry.
+	Reject metav1.ConditionStatus = "Reject"
+	// The check might pass if there was fewer workloads. Proceed
+	// with any pending workload preemption.
+	QuotaExhausted metav1.ConditionStatus = "QuotaExhausted"
+)
+
+// AdmissionCheckSpec defines the desired state of AdmissionCheck.
+type AdmissionCheckSpec struct {
+	// Name of the controller which will actually perform
+	// the checks. This is the name with which controller indentifies with,
+	// not a K8S pod or deployment name. Cannot be empty. 
+	ControllerName string `json:"controllerName"`
+
+	// How long to keep the workload suspended 
+	// after a failed check (after it transitioned to False).
+	// After that the check state goes to "Unknown".
+	// The default is 15 min. 
+	RetryDelayMinutes *int64. `json:"retryDelayMinutes,omitempty"`
+
+	// A reference to the additional parameters for the check. 
+	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`
+
+	// How should the workload book quota (and possibly preempt others) 
+	// before the check pass. The default is NoBooking.
+	// If different modes are used for different AdmissionChecks,
+	// the most restrictive one is used.
+	QuotaBooking *QuotaBookingMode `json:"quotaBooking,omitempty"`
+}
+
+const (
+	// No booking, unless a check explicitely asks for it reporting quota exhaused.
+	NoBooking QuotaBookingMode = "NoBooking"
+	
+	// Book quota (and preempt) immediately after the basic quota checks 
+	// are done, without waiting for the additional checks to cpmplete.
+	BookAfterQuotaCheck QuotaBookingMode = "BookAfterQuotaCheck"
+)
+
+// Points to where the parameters for the checks are. 
+// As clusterqueue are in cluster scope - this should be
+// a dedicated CRD specific for the Controller.
+type AdmissionCheckParametersReference struct {
+	// ApiGroup is the group for the resource being referenced. 
+	APIGroup string `json:"apiGroup"`
+
+	// Kind is the type of the resource being referenced.
+	Kind string `json:"kind"`
+
+	// Name is the name of the resource being referenced.
+	Name string `json:"name"`
+}
+```
+
+For every workload that is put to a ClusterQueue that has AdmissionChecks
+configured Kueue will add:
+
+* “AdmissionPrecheck” set false to Conditions.
```

Comment:
I would prefer to stay with Precheck - in near future we may let a bit more workloads past the Precheck than the quota would allow (and assume that the Check will effectively throttle/queue/apply quota to them).

Reply:
+1 on Marcin. The quota might be violated if we defer preemption.

Reply:
Actually, to follow https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties,
It should be AdmissionPrechecked

Reply:
From @trasc in https://github.com/kubernetes-sigs/kueue/pull/1075#discussion_r1303223834

> There are two reasons why I don't think it's a good idea:

> 1. This can imply that is done before the checks.

This is true. We want a Condition that, when true, `admissionChecks` can start.

> 2. I think, the Precheck, as described in the KEP is not blocking the quota

Correct, and that's why naming it QuotaReserved doesn't seem accurate. Unless you are suggesting having 3 conditions?
Can you elaborate?

Reply:
> This is true. We want a Condition that, when true, `admissionChecks` can start.

There is no reason to block the admissionCheck until the Quota is reserved/checked, a controller can choose to do it only after, but i don't see any value forcing this.

> Correct, and that's why naming it QuotaReserved doesn't seem accurate. Unless you are suggesting having 3 conditions?
Can you elaborate?

In #1045 having the condition (which is still called Admitted there, #1075 only renames it) means that the quota is blocked for the workload in question. 

 Now, if there is a usecase in which a check controller needs to know that at some point in the past the workload wold have had the option to block the quota, what it could do is. 

1. Keep it's condition in a non False state (preferably Unknown)
2. The scheduler will reserve the quota at some point. 
3. The controller can switch it's condition to false. (this will cancel the quota reservation)
4. Perform it's check. And set its condition to True.
5. The workload will go into the scheduling again.

(if the controller has a quick way of doing the check in some cases, the reservation cancelling and the second scheduling can be skipped)

Reply:
Let's recap.
In implementation terms, what we need is a condition that indicates that Kueue has executed the scheduling cycle and, if necessary, found candidates for preemption.

Preemption might or might not happen at this point, depending on what is set in `quotaBooking` . If this setting says `NoBooking`, meaning that we shouldn't do preemption, then it doesn't make sense to call the condition `QuotaReserved`.

`PreCheck` is a bit more neutral, although a bit vague. Let me think about it more.

Reply:
How about `NominatedForAdmission`?
- It goes along with the fact that we fill up the `admissionField`.
- It doesn't give any indication that `admissionChecks` run before or after.
- It doesn't give any indication about whether preemption happened or quota is reserved.

Reply:
But it has no indication that the quota is blocked.